csi/controller_server.go

```diff
@@ -623,53 +708,59 @@ func (cs *ControllerServer) CreateSnapshot(ctx context.Context, req *csi.CreateS
-	rsp := createSnapshotResponse(existVol.Name, backupStatus.Id, snapshot.Created, existVol.Size, backupStatus.State)
-	logrus.Debugf("ControllerServer CreateSnapshot rsp: %v", rsp)
+	logrus.Infof("createCSISnapshotTypeLonghornBackup: volume %s backup %s of snapshot %s in progress", existVol.Name, backupStatus.Id, csiSnapshotName)
+	snapshotId := encodeSnapshotID(csiSnapshotTypeLonghornBackup, existVol.Name, backupStatus.Id)
+	rsp := createSnapshotResponse(existVol.Name, snapshotId, snapshot.Created, existVol.Size, backupStatus.State == string(longhorn.BackupStateCompleted))
 	return rsp, nil
 }
 
-func createSnapshotResponse(volumeName, backupName, snapshotTime, volumeSize string, backupState string) *csi.CreateSnapshotResponse {
+func createSnapshotResponse(sourceVolumeName, snapshotID, snapshotTime, sourceVolumeSize string, readyToUse bool) *csi.CreateSnapshotResponse {
 	creationTime, err := toProtoTimestamp(snapshotTime)
 	if err != nil {
-		logrus.Errorf("Failed to parse creation time %v for backup %v", snapshotTime, backupName)
+		logrus.Errorf("Failed to parse creation time %v for csi snapshot %v", snapshotTime, snapshotID)
 	}
 
-	size, _ := util.ConvertSize(volumeSize)
+	size, _ := util.ConvertSize(sourceVolumeSize)
 	size = util.RoundUpSize(size)
-	snapshotID := encodeSnapshotID(volumeName, backupName)
 	return &csi.CreateSnapshotResponse{
 		Snapshot: &csi.Snapshot{
 			SizeBytes:      size,
 			SnapshotId:     snapshotID,
-			SourceVolumeId: volumeName,
+			SourceVolumeId: sourceVolumeName,
 			CreationTime:   creationTime,
-			ReadyToUse:     backupState == backupStateCompleted,
+			ReadyToUse:     readyToUse,
 		},
 	}
 }
 
-// encodeSnapshotID encodes the backup volume as part of the snapshotID
-// so we don't need to iterate over all backup volumes,
-// when trying to find a backup for deletion or restoration
-func encodeSnapshotID(volumeName, backupName string) string {
-	return fmt.Sprintf("bs://%s/%s", volumeName, backupName)
+func encodeSnapshotID(csiSnapshotType, sourceVolumeName, id string) string {
+	if csiSnapshotType == csiSnapshotTypeLonghornSnapshot {
+		return fmt.Sprintf("snap://%s/%s", sourceVolumeName, id)
+	} else if csiSnapshotType == csiSnapshotTypeLonghornBackup {
+		// encodeSnapshotID encodes the backup volume as part of the snapshotID
+		// so we don't need to iterate over all backup volumes,
+		// when trying to find a backup for deletion or restoration
+		return fmt.Sprintf("bs://%s/%s", sourceVolumeName, id)
+	}
+	return "INVALID-ID"
 }
 
-// decodeSnapshotID splits up the snapshotID back into it's components
-// backupType will be used once we implement the VolumeBackup crd
-func decodeSnapshotID(snapshotID string) (backupType, volumeName, backupName string) {
+func decodeSnapshotID(snapshotID string) (csiSnapshotType, sourceVolumeName, id string) {
 	split := strings.Split(snapshotID, "://")
 	if len(split) < 2 {
 		return "", "", snapshotID
 	}
-	backupType = split[0]
+	if split[0] == "bs" {
```

Comment:
NIT: Please create constants for `snap` and `bs`. e.g., `SnapshotIDPrefixLonghornSnapshot`, `SnapshotIDPrefixLonghornBackup`

Reply:
Thank you. Fixed